lib/ocpp/v201/device_model.cpp

```diff
@@ -34,6 +34,42 @@ bool DeviceModel::component_criteria_match(const Component& component,
     return false;
 }
 
+bool DeviceModel::component_criteria_match_custom(const Component& component,
+                                                  const std::vector<ComponentCriterionEnum>& component_criteria) {
+    if (component_criteria.empty()) {
+        return false;
+    }
+    for (const auto& criteria : component_criteria) {
+        const Variable variable = {conversions::component_criterion_enum_to_string(criteria)};
+
+        const auto response = this->request_value<bool>(component, variable, AttributeEnum::Actual);
+        auto value = response.value;
+        if (response.status == GetVariableStatusEnum::Accepted and value.has_value() and value.value()) {
+            return true;
+        }
+        // also send true if the component crietria isn't part of the component except "problem"
+        else if (!value.has_value() and (variable.name != "Problem")) {
+            return true;
+        }
+    }
+    return false;
+}
+bool DeviceModel::component_variables_match(const std::vector<ComponentVariable>& component_variables,
+                                            const ocpp::v201::Component& component_,
+                                            const struct ocpp::v201::Variable& variable_) {
+
+    return std::find_if(
```

Comment:
After looking at the other functions you could still simplify this with a similar for loop and split up the conditions in multiple if statements.

If a statement is true you can do an early return true. If not then you continue looping. Returning false if you exit the loop after finding no matches.